src/Passes/Source/Rules/OperandPrototype.hpp

```diff
@@ -0,0 +1,105 @@
+    /// value should be stored under.
+    class IOperandPrototype
+    {
+      public:
+        using Instruction = llvm::Instruction;
+        using String      = std::string;
+        using Value       = llvm::Value;
+        using Child       = std::shared_ptr<IOperandPrototype>;
+        using Children    = std::vector<Child>;
+        using Captures    = std::unordered_map<std::string, Value*>;
+
+        // Constructors and destructors
+        //
+
+        IOperandPrototype() = default;
+        virtual ~IOperandPrototype();
+
+        // Interface functions
+        //
+
+        /// Interface function which determines if a given Value (and its precessors) matches the
+        /// implemented pattern. If any Values are required to be captured, they will be recorded in the
+        /// captures variable.
+        virtual bool match(Value* value, Captures& captures) const = 0;
+
+        /// Interface function which defines a copy operation of the underlying implementation. Note that
+        /// unlike normal copy operators this operation returns a shared pointer to the new copy.
+        virtual Child copy() const = 0;
+
+        // Shared functionality
+        //
+
+        /// Adds a child to be matched against the matches children. Children
+        /// are matched in order and by size.
+        void addChild(Child const& child);
+
+        /// Flags that this operand should be captured. This function ensures
+        /// that the captured operand is given a name. The subsequent logic
+        /// in this class is responsible for capturing (upon match) and
+        /// uncapturing (upon backtrack) with specified name
+        void enableCapture(std::string capture_name);
+
+      protected:
+        // Function to indicate match success or failure. Either of these
+        // must be called prior to return from an implementation of
+        // IOperandPrototype::match.
+        //
+
+        /// Function which should be called whenever a match fails.
+        bool fail(Value* value, Captures& captures) const;
+
+        /// Function which should be called whenever a match is successful.
+        bool success(Value* value, Captures& captures) const;
+
+        // Helper functions for the capture logic.
+        //
+
+        /// Subroutine to match all children.
+        bool matchChildren(Value* value, Captures& captures) const;
+
+        /// Captures the value into the captures table if needed.
+        void capture(Value* value, Captures& captures) const;
+
+        /// Removes any captures from the captures table upon backtracking
+        void uncapture(Value* value, Captures& captures) const;
+
+        // Helper functions for operation
+        //
+
+        /// Shallow copy of the operand to allow name change
+        /// of the capture
+        void copyPropertiesFrom(IOperandPrototype const& other)
+        {
+            capture_name_ = other.capture_name_;
+            children_     = other.children_;
+        }
```

Comment:
Are you sure it is a shallow copy? To me it looks identical to the fully-functional copy-assignment operator (creating a deep copy). If I'm right then you may want to consider using the copy-assignment operator instead of this function.

Reply:
Yes, it is shallow in the sense that it does not make copies of all the children (only copies of their pointers). That said, you are right that it is identical to copy-assignment _but only for the base class_. Moreover, on classes inheriting from this class, we use move semantics as default behaviour which implicitly deletes the copy assignment.

Reply:
Rather important explanation, thanks.
(you may want to save such explanations as comments in the code, so that the other readers also understand)

(Up to you) If "it is identical to copy-assignment" then why not use copy-assignment instead? (copy-assignment can be `protected` too, if you'd prefer to keep the mechanism `protected`)
```c++
protected:
  IOperandPrototype& operator=(IOperandPrototype const &) = default;
```

Reply:
There are a couple of reasons why copy assignment is not used. To understand this, let's first look at how it is used: Any operand implementation will be a shared pointer to ensure that any operand tree we build is efficiently moved around and/or inserted as a child. Intuitively, we should never have to copy any operand (as this is exactly why we chose a shared pointer to start with). However, since operands have the possibility of naming their captures, we encounter the following problem:
```c++
auto any_op = any(); //< creates a shared pointer to AnyOperand
auto pattern = call("foo", "arg1"_cap = any_op, "arg2"_cap = any_op); //< creates a shared pointer to CallOperand
```
The intention of these two lines is to create a tree-like this:
```
            ┌─────────────┐            
            │ CallOperand │            
            │ capture: "" │            
            └─────────────┘            
                   │                   
         ┌─────────┴──────────┐        
┌────────▼───────┐   ┌────────▼───────┐
│   AnyOperand   │   │   AnyOperand   │
│capture: "arg1" │   │capture: "arg2" │
└────────────────┘   └────────────────┘
```
The names of the `AnyOperand` is set by `"name"_cap = operand`. One implementation that would work (in some cases) would be translating `"name"_cap = operand` into `operand->captureAs("name"); return operand`. However, this implementation is wrong and would produce the following tree (since we are using shared pointers):
```
            ┌─────────────┐            
            │ CallOperand │            
            │ capture: "" │            
            └─────────────┘            
                   │                   
         ┌─────────┴──────────┐        
┌────────▼───────┐   ┌────────▼───────┐
│   AnyOperand   │   │   AnyOperand   │
│capture: "arg2" │   │capture: "arg2" │
└────────────────┘   └────────────────┘
```
Note `arg2` in the first node. One (correct) implementation would be to make a full copy of the tree every time we set a capture. This would be terribly inefficient, so rather the implementation only copies the node that is being assigned leaving all children pointers. Since we are dealing with shared pointers, it makes sense to have a virtual`copy` function which is responsible for performing the correct copy (remember that we only have the interface at the point where we want to perform the copy). A typical implementation of `copy` could be 
```c++
    BitCastPattern::Child BitCastPattern::copy() const
    {
        auto ret = std::make_shared<BitCastPattern>();
        ret->copyPropertiesFrom(*this);
        return std::move(ret);
    }
```
Inside this function, we could choose to use a copy assignment for copying the properties of the base class. The implementation would look something like
```c++
    BitCastPattern::Child BitCastPattern::copy() const
    {
        auto ret = std::make_shared<BitCastPattern>();
        auto &target = *static_cast<IOperandPrototype*>(ret.get());
        target = *static_cast<IOperandPrototype*>(this);
        return std::move(ret);
    }
```
While this would be a technical correct implementation, it would lower the code readability substantially. Moreover given that a copy of `IOperandProtoype` in itself does not make sense on its own (unless it is a virtual operator).

It may be that it is possible to come up with a nice pattern that would use virtual assignment operators, but even so for higher classes, copy assignment is implicitly banned as the move operator is defaulted for efficiency purposes.

I hope this clarifies some of my thinking about this specific design.